target/linux/ramips/image/lzma-loader/src/board-rt305x.c

```diff
@@ -0,0 +1,16 @@
+/*
+ * Arch specific code for RT305x based boards
+ *
+ * Copyright (C) 2023 Antonio Vázquez <antoniovazquezblanco@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation.
+ */
+
+#include <stdint.h>
+
+uint32_t volatile * const UART_BASE = (uint32_t volatile * const) 0x10000500;
+uint32_t volatile * const UART_THR = UART_BASE + 0x04;
+uint32_t volatile * const UART_LSR = UART_BASE + 0x1c;
+uint32_t const UART_LSR_THRE_MASK = 0x20;
```

Comment:
I don't like keeping the pointer declaration seperated from the source-file.

I do prefer having a single header file which uses compiler-definitions for the UART base address. All platforms emulate a 16850, so the register offsets and function are identical.

Reply:
I also don't like C file declaration of pointers... It is plainly not nice but I did it for a couple of reasons:

* SOC definitions were not available in header files unless I filled the makefiles with conditionals... I don't know if this was due to the fact that the project has a makefile to call another makefile... Will do some testing.
* Not all UARTs have the same offsets. It seems MT7621 breaks the pattern... :/

I will perfom some tests and will propose a better solution, preferably based on header definitions :).

Reply:
From what I have read, `CONFIG_` variables will not be available in lzma-loader unless passed via make calls. Thinking about this I was wondering what would be the ideal scenario and thought of the following:

What is the relevant information I need for making the UART work?
I need the UART peripheral "model" and the address where it is placed. I believe that passing that information would be a saner option than passing a `CONFIG_SOC_XXXX` macro and deriving the UART behaviour and address from that information.

I guess it may be kind of difficult to come up with UART peripheral "models"... 

I could provide a dummy (silent) default UART configuration. If an UART model is specified via a definition then corresponding C files will get compiled to provide the corresponding functionality. Peripheral base address could then be specified via another paramter.

Would this approach fit your concerns? Any cons regarding this?

Thanks!

Reply:
If there's any way you can use kernel config variables, I actually tackled the same problem https://gitlab.com/olliver/openwrt/openwrt/-/commit/23a975a79f1a4b4c4f19e552343b23061497737e :) (i'll upstream it once I'm confident in it)

Reply:
@oliv3r, your suggestion is to pass down the `CONFIG_SOC_XXX` variable from the kernel config down to the loader via Make calls? Is there a better way to let those variables be known to the loader Makefiles and source code?

Reply:
I'd be surprised if that's possible or acceptable :) Just some food for thought is all.

Reply:
Drive by comment, but all that sounds like what linux already does? grep for DEBUG_UART_

Reply:
Nice, those CONFIG_DEBUG_UART_ are really useful. A combination of `CONFIG_SOC_XXXX`, `CONFIG_DEBUG_UART_BASE` and optionally `CONFIG_DEBUG_UART_CLOCK` from the kernel config should be enough for the job...

Now onto how to get those in the loader code from the Kernel config...
Thanks!

Reply:
Currently using SUBTARGET Makefile variable. I hope this will do for the time being.

Reply:
Should I resolve this comments or should @blocktrron? What is your usual way to proceed? Thanks!